src/core/socket.c

```diff
@@ -466,13 +482,57 @@ static int socket_verify(Socket *s) {
         return 0;
 }
 
+static void socket_address_hash_func(const void *p, struct siphash *state) {
+        const SocketPeer *s = p;
+
+        assert(s);
+
+        siphash24_compress(&s->family, sizeof(s->family), state);
+
+        if (s->family == AF_INET)
+                siphash24_compress(&s->peer.in.sin_addr, sizeof(s->peer.in.sin_addr), state);
+        else if (s->family == AF_INET6)
+                siphash24_compress(&s->peer.in6.sin6_addr, sizeof(s->peer.in6.sin6_addr), state);
+}
+
+static int socket_address_compare_func(const void *a, const void *b) {
+        const SocketPeer *x = a, *y = b;
+
+        if (x->family < y->family)
+                return -1;
+        if (x->family > y->family)
+                return 1;
+
+        switch(x->family) {
+        case AF_INET:
+                return memcmp(&x->peer.in.sin_addr, &y->peer.in.sin_addr, sizeof(x->peer.in.sin_addr));
+        case AF_INET6:
+                return memcmp(&x->peer.in6.sin6_addr, &y->peer.in6.sin6_addr, sizeof(x->peer.in6.sin6_addr));
+        }
+
+        return -1;
```

Comment:
ah, actually @ssahani is right, this is a "union socket_addr_union", not a "union in_addr_union", hence doing my memcmp() suggestion doesn't really work.

That said, @ssahani I think it would make sense to rework this so that SocketPeer only stores an "union in_addr_union" instead of a full socket address, after all we don't know what to do with non-IP addresses anyway, and don#t care for port numbers either.